orttraining/orttraining/python/training/orttrainer.py

```diff
@@ -843,3 +854,306 @@ def _update_onnx_model_initializers(self, state_tensors):
         for w_i in replace_indices:
             del self._onnx_model.graph.initializer[w_i]
         self._onnx_model.graph.initializer.extend(new_weights)
+
+    def _extract_model_states(self, states):
+        """Extract model states from the training session and load into the states dictionary"""
+
+        model = 'model'
+        model_states = self._training_session.get_model_state(include_mixed_precision_weights=False)
+        states[model] = {}
+
+        # extract trainer model weights from the training session
+        for precision in model_states:
+            states[model][precision] = {}
+            for model_state_key in model_states[precision]:
+                states[model][precision][model_state_key] = torch.from_numpy(model_states[precision][model_state_key])
+
+        # extract untrained (frozen) model weights
+        precision_map = {TensorProto.FLOAT: 'fp32'}
+        for node in self._onnx_model.graph.initializer:
+            if node.name not in states[model][precision] and node.name in self.options.utils.frozen_weights:
+                if node.data_type in precision_map:
+                    states[model][precision_map[node.data_type]][node.name] = \
+                        torch.from_numpy(np.array(onnx.numpy_helper.to_array(node)))
```

Comment:
is there a reason for returning the checkpoint in torch tensor format but converting i to numpy while loading? Is it possible to stick to numpy?

Reply:
I am not sure on this. I was trying to keep it consistent with what we had in the existing design.
The idea is that users want to use the state dict to load into PyTorch. If PyTorch can accept numpy arrays, then we can only work with numpy and not depend on converting to tensors. Do you know if ```pytorch_model.load_state_dict()``` can load from numpy arrays instead of tensors?

Reply:
I think we can leave this as is (returning a torch tensor). to keep it in sync with the existing implementation unless there is a strong reason to return it as a numpy array instead.

Reply:
we could potentially gain some perf by skipping the extra copy. The relevant torch code comment states only limiting to persistent buffer so it should by okay (https://github.com/pytorch/pytorch/blob/80f7510d9202bc46fa3faeb798ba0d7f6db2f67a/torch/nn/modules/module.py#L1019). I think we have some distributed tests that load the state_dict returned by ORT into the pytorch model. We can give it a try and verify on those tests.

Reply:
It appears that torch ```load_state_dict``` throws an error if numpy array is provided instead of torch tensors. The two solution I can think of:
1. Return torch tensors when ```pytorch_format``` is true and numpy arrays otherwise.
2. Always return a torch tensor and do the conversion ourselves from tensor to numpy.

I am leaning more towards the first one. Will make the changes. Let me know if any concerns.